### PR TITLE
feat: add proxy rotation (settable per store)

### PR DESCRIPTION
**Pull request by Doridian**
### Description

Adds ability to add multiple proxies per store, rotated once per fetch.
Proxies are read from a file `STORENAME.proxies` in the format like `socks5://username:password@ip` one per line.
Adds some require architectural changes for that to ensure adblocker doesn't explode (puppeteer only supports one request listener per instance, so we have to write our own chain there)

### CODE REVIEW

**Comment by jef** [comment]
@Doridian with the great contribution! Thanks for help. I'll take a look at this tonight and get back.

**Comment by MujiSayed** [comment]
quick question about this. I know format is  socks5://username:password@ip, but what if there is no username/password for said socks5 proxies? Also since default port is 1080, is that what it's guessing as it's port or will a format of socks5://ip:port work as well?

Thanks for the contribution!

**Comment by Doridian** [comment]
Correct. That format should work. Default port is 1080 indeed.

File: src/store/lookup.ts [hunk 1]
```diff
@@ -9,18 +9,109 @@ import {
...
+
+	return store.proxyList[store.currentProxyIndex];
+}
+
+async function handleLowBandwidth(request: Request) {
+	if (!config.browser.lowBandwidth) {
+		return false;
+	}
+
+	const typ = request.resourceType();
```
Comment (jef) [inline src/store/lookup.ts:51]: Is this because `type` is a reserved word? Haha :stuck_out_tongue_closed_eyes:

**Review by jef — APPROVED** [review]
Amazing. Thank you so much @Doridian. This really needed some clean up. I appreciate the time spent on the refactoring as well!

**Comment by Doridian** [comment]
Indeed it is. 

On Tue, Dec 1, 2020, at 16:11, Jef LeCompte wrote:
> 

> *@jef* approved this pull request.

> Amazing. Thank you so much @Doridian <https://github.com/Doridian>. This really needed some clean up. I appreciate the time spent on the refactoring as well!

> 
> In src/store/lookup.ts <https://github.com/jef/streetmerchant/pull/1026#discussion_r533806329>:

> > +		store.currentProxyIndex = 0;
+	}
+
+	logger.info(
+		`ℹ [${store.name}] Next proxy index: ${store.currentProxyIndex} / Count: ${store.proxyList.length}`
+	);
+
+	return store.proxyList[store.currentProxyIndex];
+}
+
+async function handleLowBandwidth(request: Request) {
+	if (!config.browser.lowBandwidth) {
+		return false;
+	}
+
+	const typ = request.resourceType();

**Comment by JasonMore** [comment]
@Doridian which socks5 proxies do you use? Looking for suggestions

**Comment by gigi2006** [comment]
@Doridian @jef 
would also be very interested which proxys you recommend...

**Pull request merged**